onnxruntime/core/providers/qnn/builder/qnn_model.cc

```diff
@@ -199,20 +200,41 @@ static Status BindQnnTensorMemoryToOrtValueMemory(const logging::Logger& logger,
                                                   void* ort_value_data, uint32_t ort_value_data_size,
                                                   Qnn_ContextHandle_t qnn_context,
                                                   Qnn_Tensor_t& qnn_tensor) {
-  // either set qnn_tensor memHandle or clientBuf
-  const bool uses_shared_memory = ort_value_memory_info == HtpSharedMemoryAllocator::AssociatedMemoryInfo();
 
-  if (!uses_shared_memory) {
-    LOGS(logger, VERBOSE) << "Setting Qnn_Tensor_t clientBuf to ORT tensor memory.";
-    SetQnnTensorMemType(qnn_tensor, QNN_TENSORMEMTYPE_RAW);
-    SetQnnTensorClientBuf(qnn_tensor, ort_value_data, ort_value_data_size);
-  } else {
+  const bool uses_shared_mem_allocator = ort_value_memory_info == HtpSharedMemoryAllocator::AssociatedMemoryInfo();
+  const bool buffer_aligned = ((uintptr_t)ort_value_data % rpcmem::BUFFER_ALIGNMENT_BLOCK_SIZE) == 0
+                             && ort_value_data_size % rpcmem::BUFFER_ALIGNMENT_BLOCK_SIZE == 0;
+  const bool uses_htp_backend = qnn_backend_manager.GetQnnBackendType() == QnnBackendType::HTP;
+
+  // if HTP backend is in use and the buffer pointer is CPU allocated and 4K aligned,
+  // the buffer needs to be mapped to RPC memory space and registered
+  if (uses_shared_mem_allocator || (uses_htp_backend && buffer_aligned)) {
     LOGS(logger, VERBOSE) << "Setting Qnn_Tensor_t memHandle to ORT tensor shared memory.";
-    Qnn_MemHandle_t qnn_mem_handle{};
+    Qnn_MemHandle_t qnn_mem_handle;
     ORT_RETURN_IF_ERROR(qnn_backend_manager.GetOrRegisterContextMemHandle(qnn_context, ort_value_data, qnn_tensor,
-                                                                          qnn_mem_handle));
+                                                                          qnn_mem_handle, uses_shared_mem_allocator));
     SetQnnTensorMemType(qnn_tensor, QNN_TENSORMEMTYPE_MEMHANDLE);
     SetQnnTensorMemHandle(qnn_tensor, qnn_mem_handle);
+
+  } else {
+    LOGS(logger, VERBOSE) << "Setting Qnn_Tensor_t clientBuf to ORT tensor memory.";
+    SetQnnTensorMemType(qnn_tensor, QNN_TENSORMEMTYPE_RAW);
+    SetQnnTensorClientBuf(qnn_tensor, ort_value_data, ort_value_data_size);
+  }
+
+  return Status::OK();
+}
+
+static Status UnregisterQnnTensorMemory(QnnBackendManager& qnn_backend_manager,
+                                        const OrtMemoryInfo& ort_value_memory_info,
+                                        void* ort_value_data, uint32_t ort_value_data_size,
+                                        Qnn_ContextHandle_t qnn_context) {
+  const bool uses_shared_mem_alloc = ort_value_memory_info == HtpSharedMemoryAllocator::AssociatedMemoryInfo();
+  const bool buffer_aligned = ((uintptr_t)ort_value_data % rpcmem::BUFFER_ALIGNMENT_BLOCK_SIZE) == 0 && ort_value_data_size % rpcmem::BUFFER_ALIGNMENT_BLOCK_SIZE == 0;
```

Comment:
nit: `reinterpret_cast<uintptr_t>(ort_value_data)` and line length. maybe put this condition in a helper function?